app/src/main/java/tech/relaycorp/gateway/background/endpoint/EndpointPreRegistrationService.kt

```diff
@@ -0,0 +1,55 @@
+package tech.relaycorp.gateway.background.endpoint
+
+import android.app.Service
+import android.content.Intent
+import android.os.Handler
+import android.os.IBinder
+import android.os.Looper
+import android.os.Message
+import android.os.Messenger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.runBlocking
+import tech.relaycorp.gateway.background.component
+import tech.relaycorp.gateway.domain.endpoint.EndpointRegistration
+import javax.inject.Inject
+
+class EndpointPreRegistrationService : Service() {
+    private val scope get() = CoroutineScope(Dispatchers.IO)
+
+    @Inject
+    lateinit var endpointRegistration: EndpointRegistration
+
+    override fun onBind(intent: Intent): IBinder? {
+        component.inject(this)
+        val messenger = Messenger(
+            object : Handler(Looper.getMainLooper()) {
+                override fun handleMessage(msg: Message) {
+                    when (msg.what) {
+                        PREREGISTRATION_REQUEST -> reply(msg)
+                    }
+                }
+            }
+        )
+        return messenger.binder
+    }
+
+    // TODO: Replace runBlocking with scope.launch
+    internal fun reply(requestMessage: Message) = runBlocking {
```

Comment:
I was trying to use `scope.launch` but the instrumentation test kept failing. Should I be doing something else here or in the test?

Reply:
> In the test, you would need some way to control the scope, either by overriding it, or receiving it through the constructor

How can I do that? I guess I'd have to change something here but I don't know how:

https://github.com/relaycorp/relaynet-gateway-android/blob/b551c056722c36a8a5947b190230a940df3d8b23/app/src/androidTest/java/tech/relaycorp/gateway/background/endpoint/EndpointPreRegistrationServiceTest.kt#L39-L43

> Or you need to make sure the test is runBlocking(Dispatchers.IO) {} I believe

I changed the `runBlocking` to `scope.launch` in the `reply()` and then changed the test to pass `Dispatchers.IO` to `runBlocking`, but now I get the following error:

```
08/11 16:49:14: Launching 'requestPreRegistra...()' on Pixel 2 API 29.
Running tests

$ adb shell am instrument -w -r    -e debug false -e class 'tech.relaycorp.gateway.background.endpoint.EndpointPreRegistrationServiceTest#requestPreRegistration' tech.relaycorp.gateway.test/tech.relaycorp.gateway.test.TestAndroidJUnitRunner
Connected to process 12367 on device 'Pixel_2_API_29 [emulator-5554]'.

Started running tests


Test failed to run to completion. Reason: 'Instrumentation run failed due to 'Process crashed.''. Check device logcat for details
Test running failed: Instrumentation run failed due to 'Process crashed.'
Tests ran to completion.
```

Logcat:

```
2020-08-11 16:52:31.632 12940-12977/tech.relaycorp.gateway E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: tech.relaycorp.gateway, PID: 12940
    java.lang.NullPointerException: Attempt to invoke virtual method 'void android.os.Messenger.send(android.os.Message)' on a null object reference
        at tech.relaycorp.gateway.background.endpoint.EndpointPreRegistrationService$reply$1.invokeSuspend(EndpointPreRegistrationService.kt:43)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```